feat: add recipe details and comments

pantpad opened the pull request:
This PR is linked to PR#15

### Challenge #4: Recipe Details and Comments

Develop a recipe details page where users can view the full recipe, including ingredients, instructions, and user
comments. 
Enable users to add comments and rate the recipe, displaying the average rating and updating the list of
comments.

pantpad committed:
chore: add comment on recipe details

pantpad committed:
feat: add all the recipe details + form to submit review

pantpad committed:
feat: add custom submit function to handle form comment submission

pantpad committed:
feat: make the data appear as european format + add local state for added comments

pantpad committed:
refactor: move recipe comment into its own component

pantpad committed:
feat: reset comment form on correct submit

pantpad committed:
chore: add form comments placeholders

francescomarins commented on frontend/src/api/recipe.ts:
```diff
@@ -49,4 +57,21 @@ export const submitRecipe = async (formData: FormData) => {
+    recipeId: string = ''
+) => {
+    const formObject = Object.fromEntries(formData.entries())
+    const formWithData = Object.assign({}, formObject, { date: new Date() })
```
This could be simplified using the spread syntax: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

Have a look here ;)

francescomarins commented on frontend/src/api/recipe.ts:
```diff
@@ -49,4 +57,21 @@ export const submitRecipe = async (formData: FormData) => {
 
+export const submitComment = async (
+    formData: FormData,
+    recipeId: string = ''
```
In TypeScript, when you declare a parameter with a default value (like recipeId: string = ''), it makes the parameter optional in terms of function invocation, even though the syntax doesn't include a ? mark.

Why using a default value? This parameter is required. Removing it ensures you to use the function properly when ivoking it

francescomarins commented on frontend/src/components/Recipes/Details/RecipeComment.tsx:
```diff
@@ -0,0 +1,20 @@
+        <div className="rounded border bg-white p-4 shadow">
+            <p>{comment}</p>
+            <p>
+                {new Date(date).toLocaleDateString('de-DE', {
```
I’m really curious about how the idea of using the German locale came about ❤️😂

francescomarins commented on frontend/src/components/Recipes/Details/RecipeComments.tsx:
```diff
@@ -0,0 +1,108 @@
+}
+
+export function RecipeComments({ recipeId }: RecipeCommentsProps) {
+    const formRef = useRef<HTMLFormElement>(null)
```
Personally, I would have used a controlled form, have you chosen an uncontrolled approach for any specific reason?

pantpad replied:
Want to try having 2 different form types when introducing react-hook-forms.

francescomarins commented on frontend/src/components/Recipes/Details/RecipeComments.tsx:
```diff
@@ -0,0 +1,108 @@
+    )
+
+    const [recipeAddedComments, setRecipeAddedComments] = useState(
+        [] as RecipeCommentType[]
```
You could use the useState<RecipeCommentType[]> notation

francescomarins commented on frontend/src/components/Recipes/Details/RecipeComments.tsx:
```diff
@@ -0,0 +1,108 @@
+            ))}
+            {recipeAddedComments.map((comment) => (
+                <RecipeComment key={comment.id} {...comment} />
+            ))}
```
You may put it below the form so you avoid having layout shift after adding a comment, what do you think?

pantpad committed:
feat: address pr15 comments

make comments display under the textarea
sort comments by most recent
remove useRef on form
remove german locale, use undefined
improve typing on recipeId and checks if undefined
use spread operator to create correct formWithDate
